### PR TITLE
Service provider interface

### PR DESCRIPTION
**Pull request by burzum**

### CODE REVIEW

File: src/AuthenticationServiceProviderInterface.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ *  AuthenticationServiceProviderInterface
+ */
+interface AuthenticationServiceProviderInterface
+{
+    /**
+     * Returns a service provider instance.
```
Comment (robertpustulka) [inline src/AuthenticationServiceProviderInterface.php]: Returns an authentication service, not a provider.

File: tests/TestCase/Middleware/AuthenticationMiddlewareTest.php [hunk 2]
```diff
@@ -160,13 +136,12 @@ public function testApplicationAuthenticationRequestResponse()
 
         $application = $this->getMockBuilder(BaseApplication::class)
```
Comment (robertpustulka) [inline tests/TestCase/Middleware/AuthenticationMiddlewareTest.php]: BaseApplication does not implement provider interface. Should it be `TesApp\Application` here?

Comment (burzum) [reply tests/TestCase/Middleware/AuthenticationMiddlewareTest.php]: Change this and also tried to mock the interface, I'm always getting `1) Authentication\Test\TestCase\Middleware\AuthenticationMiddlewareTest::testApplicationAuthenticationRequestResponse
InvalidArgumentException: Subject must be an instance of `Authentication\AuthenticationServiceInterface` or `Cake\Core\HttpApplicationInterface`, `Mock_Application_985fb108` given.` this.

**Comment by robertpustulka** [comment]
Should we provide something similar for the Authorization plugin?

**Comment by burzum** [comment]
Yes

**Pull request merged**